tensorflow/stream_executor/dnn.h

```diff
@@ -2383,6 +2402,40 @@ class DnnSupport {
     return false;
   }
 
+  // Enqueue a CTC Loss operation onto the stream.
+  //
+  // Arguments:
+  //  stream: pointer to the stream where this operation should be enqueued to.
+  //  probs_desc: specifies the shape and the data layout of the input tensor.
+  //  probs_data: the device memory region that contains the input tensor.
+  //  labels_data: the device memory region that contains the labels_value
+  //    tensor.
+  //  labels_lengths_data: the device memory region that contains the
+  //    labels_lengths tensor
+  //  input_lengths_data: the device memory region that contains the seq_lengths
+  //    tensor
+  //  costs_data: the device memory region that contains the costs tensor.
+  //  grads_desc: specifies the shape and the data layout of the grads tensor.
+  //  grads_data: the device memory region that contains the grads tensor.
+  //  ctc_loss_desc: a CTCLoss descriptor created by createCTCLossDescriptor.
+  //  workspace_allocator: a memory allocator that creates the temporary
+  //    workspace memory used by this operation. The caller is responsible for
+  //    keeping the memory alive long enough for this operation, and recylces
+  //    afterwards.
+  virtual bool DoCtcLoss(Stream* stream,
```

Comment:
Can we have something like:
```
virtual bool DoCtcLoss(
  Stream* stream,
  dnn::DataType element_type,
  const dnn::RnnStateTensorDescriptor &probs_desc,
  DeviceMemoryBase probs_data,
  absl::Span<const int32> labels_data,
  absl::Span<const int32> labels_lengths_data,
  absl::Span<const int32> input_lengths_data,
  DeviceMemoryBase costs_data,
  const dnn::RnnStateTensorDescriptor &grads_desc,
  DeviceMemoryBase grads_data,
  const dnn::CtcLossDescriptor &ctc_loss_desc,
  DeviceMemoryBase scratch_memory);
```
? A few points:
* Pass DeviceMemory and Span by value. They are already pointer types, therefore no need to pass by references.
* Generalize on element type, so that one function can support multiple element types, instead of just float. See PrepareForConvolution, the virtual DoConvolve, and the class comment of DnnSupport.
* Separate scratch memory allocation from this function. Users should allocate the scratch memory separately. This is also consistent with cuDNN's API.

Reply:
For the scratch memory allocation, I follow the way used in DoBatchNormalizationForward/Backward and DoRNNForward/Backward. Shouldn't it be consistent with them?
Do I need to generalize the element type in this PR, considering the current cuDNN only support float?

Reply:
> Please refer to the comment for class DnnSupport, and PrepareForConvolution.

The file itself is already inconsistent. It was way too verbose and I simplified Convolution-related APIs. In the spirit of being as thin of a wrapper as possible (aka simplifying DnnSupport itself), we want to be close to cuDNN / MIOpen without unnecessary cleverness. Please also refer to the comments for DnnSupport class.

> Do I need to generalize the element type in this PR, considering the current cuDNN only support float?

I'd argue it's easier to implement and maintain by adding the element type. The key point is that the element type is just passed down to cuDNN without being much inspected by StreamExecutor.

For example, StreamExecutor doesn't have to know which types are supported vs which are not (because that may change with a different version of cuDNN). If a type is not supported, cuDNN will generate an error, and StreamExecutor just forwards it transparently.